### PR TITLE
Add `sret` attribute for entry point function if needed

### PR DESCRIPTION
**Pull request by 1tnguyen**
### Description

The root cause is ABI differences b/w ARM and x86 as explained [here](https://github.com/NVIDIA/cuda-quantum/issues/553#issuecomment-1800311727).

Hence, this PR adds the `sret` attribute to the first (hidden) argument if the entry point function is returning one and re-enables the `auto_kernel.cpp` test on ARM. 

Resolved: https://github.com/NVIDIA/cuda-quantum/issues/553

### CODE REVIEW

File: test/NVQPP/auto_kernel.cpp [hunk 1]
```diff
@@ -9,7 +9,6 @@
 // FIXME: enable this for all architectures,
 // see https://github.com/NVIDIA/cuda-quantum/issues/553.
```
Comment (bmhowe23) [inline test/NVQPP/auto_kernel.cpp:11]:
```suggestion
```

**Review by bmhowe23 — APPROVED** [review]
Nice.

File: test/NVQPP/auto_kernel.cpp (hunk 1 above)
Comment (1tnguyen) [reply test/NVQPP/auto_kernel.cpp]: Good catch! Somehow, I missed it :)

File: lib/Optimizer/Transforms/GenKernelExecution.cpp [hunk 2]
```diff
@@ -512,6 +512,18 @@ class GenerateKernelExecution
...
         builder.create<func::FuncOp>(loc, mangledAttr.getValue(), newFuncTy);
+    if (cudaq::opt::factory::hasHiddenSRet(funcTy)) {
+      // The first argument should be a pointer type if this function has a
+      // hidden sret.
+      if (auto ptrTy = dyn_cast<cudaq::cc::PointerType>(
+              rewriteEntry.getFunctionType().getInput(0))) {
+        auto eleTy = ptrTy.getElementType();
+        rewriteEntry.setArgAttr(0,
+                                mlir::LLVM::LLVMDialect::getStructRetAttrName(),
+                                mlir::TypeAttr::get(eleTy));
```
Comment (schweitzpgi) [inline lib/Optimizer/Transforms/GenKernelExecution.cpp:523]: nit: Don't need the `mlir::`

**Pull request merged**